dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java

```diff
@@ -605,13 +608,54 @@ private boolean process(Constraint c) {
     }
 
     /*
-        This pass rewrites the memory model so that no positive constraints appear on the rhs-side of a difference.
-        Suppose "r = a \ b" where r and a are negative and b is positive.
+        This pass rewrites the memory model so that no non-negative constraints appear on the RHS of a difference.
+        Consider "r = a \ b" where b is non-negative (note that r and a must be non-positive then).
         For example, this is the case if there is a constraint "~empty(r)"
         We rewrite the model to
-            let b`= free()
+            let b'= free()
             let r = a \ b' // b' gets cut, but it is effectively a base relation so this causes no large encoding.
-            empty (r & b) // b is now in a positive position and thus not need to get cut
+            empty (r & b) // b is now in a positive position and thus not need to get cut.
+
+        We claim that this model is equisatisfiable to the original memory model.
+        To understand this intuitively, it is helpful so think of the fresh b' as the original b but with "some slack".
+        We will argue that the best choice for b' is precisely b.
+        Any larger choice for b' will make r smaller, but since r is non-positive this reduces the chances
+        of satisfying negated axioms (e.g., if "~empty(r)" is true for some small r, then it is also true for all larger r).
+        On the other hand, if b' is chosen too small, then r gets so large that it violates "empty(r & b)".
+        We prove this formally:
+
+        "NEW => ORIGINAL":
+        Suppose M' is a consistent execution of the new memory model.
+        Then from "let r = a \ b'" it follows that M'(r) \subseteq M'(a).
+        Furthermore, from "empty (r & b)" it follows that M'(r) \subseteq ~M'(b).
+        Combining both, we get M'(r) \subseteq M'(a) & ~M'(b) = M'(a) \ M'(b)
+        Now consider an execution M of the original memory model that matches with M' on the base relations
+        (modulo b' which does not exist in the original model).
+        Then M(a) = M'(a) and M(b) = M'(b), so that M'(r) \subseteq M'(a) \ M'(b) = M(a) \ M(b) = M(r).
```

Comment:
"Then M(a) = M'(a) and M(b) = M'(b)" -> why? I don't think you are assuming `a` and `b` to be base

Reply:
This is because if you know all base relations, then you also know all derived relations.
In particular, if `M` and `M'` agree on the base relations they must also agree on derived relations.
The only exception is for derived relations that depend on `b'` because this one does not appear in `M` and so they "disagree" there. However, neither `a` nor `b` depend on `b'` and so both executions agree on those.
Put differently, both executions must agree up to the point where I do the rewriting and only from there on there can be a difference.